VarNamedTuple, with an application for FastLDF (#1150)
TuringLang/DynamicPPL.jl: docs/src/internals/varnamedtuple.md

In a `VarNamedTuple` each level of such nesting of `PropertyLens`es corresponds to a level of nested `NamedTuple`s, with the `Symbol`s of the lenses as keys.
For instance, the `VarNamedTuple` mapping `@varname(x) => 1, @varname(y.z) => 2` would be stored as

```
VarNamedTuple(; x=1, y=VarNamedTuple(; z=2))
```

where `VarNamedTuple(; x=a, y=b)` is just a thin wrapper around the `NamedTuple` `(; x=a, y=b)`.

It's often handy to think of this as a tree, with each node being a `VarNamedTuple`, like so:

```
   VNT
x /   \ y
 1     VNT
         \ z
          2
```

If all `VarName`s consisted of only `PropertyLens`es we would be done designing the data structure.
However, recall that `VarName`s allow three different kinds of lenses: `PropertyLens`es, `IndexLens`es, and `identity` (the trivial lens).
The `identity` lens presents no complications, and in fact in the above example there was an implicit identity lens in e.g. `@varname(x) => 1`.
It is the `IndexLenses` that require more structure.

An `IndexLens` is the square bracket indexing part in `VarName`s like `@varname(x[1])`, `@varname(x[1].a.b[2:3])` and `@varname(x[:].b[1,2,3].c[1:5,:])`.

> `VarNamedTuple` can not deal with `IndexLens`es in their full generality, for reasons we'll discuss below.

Reviewer: ```suggestion
`VarNamedTuple` cannot deal with `IndexLens`es in their full generality, for reasons we'll discuss below.
```

julia> print(vnt)
VarNamedTuple(; a=1.0, b=VarNamedTuple(; c=[2.0, 3.0]), d=VarNamedTuple(; e=PartialArray{VarNamedTuple{(:f,), Tuple{DynamicPPL.VarNamedTuples.PartialArray{Symbol, 1}}},1}((2,) => VarNamedTuple(; f=PartialArray{Symbol,1}((3,) => hip, (4,) => hop)))))
```

The output there may be a bit hard bit hard to parse, so to illustrate:

```julia
julia> vnt[@varname(b)]
VarNamedTuple(; c=[2.0, 3.0])

julia> vnt[@varname(b.c[1])]
2.0

julia> vnt[@varname(d.e)]
PartialArray{VarNamedTuple{(:f,), Tuple{DynamicPPL.VarNamedTuples.PartialArray{Symbol, 1}}},1}((2,) => VarNamedTuple(; f=PartialArray{Symbol,1}((3,) => hip, (4,) => hop)))

julia> vnt[@varname(d.e[2].f)]
PartialArray{Symbol,1}((3,) => hip, (4,) => hop)
```

Or as a tree drawing, where `PA` marks a `PartialArray`:

```
   /----VNT------\
a /      | b      \ d
 1   [2.0, 3.0]   VNT
                   | e
                  PA(2 => VNT)
                           | f
                          PA(3 => :hip, 4 => :hop)
```

The above code also highlights how setting indices in a `VarNamedTuple` is done using `BangBang.setindex!!`.
We do not define a method for `Base.setindex!` at all, `setindex!!` is the only way.

> This is because `VarNamedTuple` mixes mutable an immutable data structures.

Reviewer: ```suggestion
This is because `VarNamedTuple` mixes mutable and immutable data structures.
```
